nf_core/pipelines/downloads/singularity.py

```diff
@@ -0,0 +1,573 @@
+            if dir and os.path.exists(dir):
+                return dir
+
+        library_dir = check_env_var("NXF_SINGULARITY_LIBRARYDIR")
+        cache_dir = check_env_var("NXF_SINGULARITY_CACHEDIR")
+
+        # Check each container in the list and defer actions
+        containers_download: List[Tuple[str, str]] = []
+        containers_pull: List[Tuple[str, str]] = []
+        containers_copy: List[Tuple[str, str, str]] = []
+
+        # We may add more tasks as containers need to be copied between the various caches
+        total_tasks = len(containers)
+
+        for container in containers:
+            container_filename = self.get_container_filename(container)
+
+            # Files in the remote cache are already downloaded and can be ignored
+            if container_filename in exclude_list:
+                log.debug(f"Skipping download of container '{container_filename}' as it is cached remotely.")
+                self.progress.update(task, advance=1, description=f"Skipping {container_filename}")
+                continue
+
+            # Generate file paths for all three locations
+            output_path = os.path.join(output_dir, container_filename)
+
+            if os.path.exists(output_path):
+                log.debug(f"Skipping download of container '{container_filename}' as it is in already present.")
+                self.progress.update(task, advance=1, description=f"{container_filename} exists at destination")
+                continue
+
+            library_path = os.path.join(library_dir, container_filename) if library_dir else None
+            cache_path = os.path.join(cache_dir, container_filename) if cache_dir else None
+
+            # get the container from the library
+            if library_path and os.path.exists(library_path):
+                containers_copy.append((container, library_path, output_path))
+                # update the cache if needed
+                if cache_path and amend_cachedir and not os.path.exists(cache_path):
+                    containers_copy.append((container, library_path, cache_path))
+                    total_tasks += 1
+                    self.progress.update(task, total=total_tasks)
+
+            # get the container from the cache
+            elif cache_path and os.path.exists(cache_path):
+                containers_copy.append((container, cache_path, output_path))
+
+            # no library or cache
+            else:
+                # fetch method (download or pull)
+                if container.startswith("http"):
+                    fetch_list = containers_download
+                else:
+                    fetch_list = containers_pull
+
+                if cache_path and amend_cachedir:
+                    # download into the cache
+                    fetch_list.append((container, cache_path))
+                    # and copy from the cache to the output
+                    containers_copy.append((container, cache_path, output_path))
```

Comment:
Why a copy to output_path? But it is already late, so it is probably me...

Reply:
This resonates with a comment you made in the previous PR. I need to double-check all those copy operations. Make sure that the "amend" / "copy" are respected, and avoid copying files if not needed